capa/main.py

```diff
@@ -773,62 +893,124 @@ def get_signatures(sigs_path: Path) -> List[Path]:
     return paths
 
 
+def get_sample_analysis(format_, arch, os_, extractor, rules_path, counts):
+    if isinstance(extractor, StaticFeatureExtractor):
+        return rdoc.StaticAnalysis(
+            format=format_,
+            arch=arch,
+            os=os_,
+            extractor=extractor.__class__.__name__,
+            rules=tuple(rules_path),
+            base_address=frz.Address.from_capa(extractor.get_base_address()),
+            layout=rdoc.StaticLayout(
+                functions=(),
+                # this is updated after capabilities have been collected.
+                # will look like:
+                #
+                # "functions": { 0x401000: { "matched_basic_blocks": [ 0x401000, 0x401005, ... ] }, ... }
+            ),
+            feature_counts=counts["feature_counts"],
+            library_functions=counts["library_functions"],
+        )
+    elif isinstance(extractor, DynamicFeatureExtractor):
+        return rdoc.DynamicAnalysis(
+            format=format_,
+            arch=arch,
+            os=os_,
+            extractor=extractor.__class__.__name__,
+            rules=tuple(rules_path),
+            layout=rdoc.DynamicLayout(
+                processes=(),
+            ),
+            feature_counts=counts["feature_counts"],
+        )
+    else:
+        raise ValueError("invalid extractor type")
+
+
 def collect_metadata(
     argv: List[str],
     sample_path: Path,
     format_: str,
     os_: str,
     rules_path: List[Path],
     extractor: FeatureExtractor,
+    counts: dict,
 ) -> rdoc.Metadata:
-    md5 = hashlib.md5()
-    sha1 = hashlib.sha1()
-    sha256 = hashlib.sha256()
-
-    assert isinstance(extractor, StaticFeatureExtractor)
-    buf = sample_path.read_bytes()
-
-    md5.update(buf)
-    sha1.update(buf)
-    sha256.update(buf)
+    # if it's a binary sample we hash it, if it's a report
+    # we fetch the hashes from the report
+    md5, sha1, sha256 = extractor.get_sample_hashes()
```

Comment:
I guess the SampleHashes iterator facilitates this, but it's a little opaque and we could also do `hashes = get_sample_hashes()` and then `hashes.md5`, etc.